### PR TITLE
chore: fix fetchIntermediateIndex, observation filling and add test

### PR DESCRIPTION
**Pull request by PangZhi**

### CODE REVIEW

File: contracts/VolOracle.sol [hunk 1]
```diff
@@ -59,24 +61,45 @@ contract VolOracle {
...
-        // both inclusive
-        if (latestPoolObservationIndex > lastObservationIndex)
-            return (lastObservationIndex + 1, latestPoolObservationIndex);
-        return (lastObservationIndex + 1, latestPoolObservationIndex + poolCardinality);
+        IUniswapV3Pool uniPool = IUniswapV3Pool(_pool);
+        (, , uint16 latestPoolObservationIndex, uint16 poolCardinality, , , ) = uniPool.slot0();
+
+        uint256 oldestObservationIndex = (latestPoolObservationIndex + 1) % poolCardinality;
+
+        (uint32 oldestObservationTs, , , bool initialized) = uniPool.observations(oldestObservationIndex);
```
Comment (ankitchiplunkar) [inline contracts/VolOracle.sol:77]: add this check as well: https://github.com/Uniswap/v3-periphery/blob/main/contracts/libraries/OracleLibrary.sol#L83

File: test/voloracle.test.ts [hunk 2]
```diff
@@ -149,5 +154,254 @@ describe("Vol Oracle tests", () => {
...
+    let startTs: number;
+
+    /**
+     * fake slot0 and observations return for uniswap. ≈
+     * @param firstIndex the index of the first observation to be mocked out
+     * @param length the length of the observations to be mocked out
+     * @param firstTs the timestamp of the first mocked out observation, increment ts by 1 per observation
+     * @param tickCumulativeData optional, if not specified, filled with tick = 1
+     */
+    async function fakeObservations(
```
Comment (ankitchiplunkar) [inline test/voloracle.test.ts]: can we put these functions in a helper.ts?

File: test/voloracle.test.ts [hunk 3]
```diff
@@ -149,5 +154,254 @@ describe("Vol Oracle tests", () => {
...
+      const observation = await volOracle.getObservation(uniV3Pool.address, observationGrowth);
+      expect(observation.blockTimestamp).to.equal(startTs + tsOffset + observationGrowth - 1);
+      expect(observation.tickCumulative).to.equal(tickCumulativeData[3]);
+      // (tsOffset - mockObservationIndex0) * 1 + 4 + 9 + 16
+      expect(observation.tickSquareCumulative).to.equal((tsOffset - mockObservationIndex0) * 1 + 4 + 9 + 16);
+    });
   });
 });
+
+async function getLatestTimestamp() {
```
Comment (ankitchiplunkar) [inline test/voloracle.test.ts]: can we put these functions in a helper.ts?

File: test/voloracle.test.ts [hunk 4]
```diff
@@ -149,5 +154,254 @@ describe("Vol Oracle tests", () => {
...
+        ...{
+          observationIndex: (firstIndex + length - 1) % observationCardinality,
+          observationCardinality: observationCardinality,
+        },
+      });
+
+      for (let i = 0; i < length; i++) {
+        const ts = firstTs + i;
+        const tickCumulative =
+          typeof tickCumulativeData !== "undefined" ? tickCumulativeData[i] : startTickCumulative + ts - startTs;
```
Comment (ankitchiplunkar) [inline test/voloracle.test.ts]: this means tick is always 1 in tests

Comment (PangZhi) [reply test/voloracle.test.ts]: this is the default value, can be overridden if the data array is passed in

**Pull request merged**